### PR TITLE
[WIP] Prototype for JSON deserialization

### PR DESCRIPTION
**Pull request by EloiZalczer**
This is a work in progress for the JSON deserialization. I would like to know your opinion, whether this is worth continuing or not.

Currently supports ListField of ReferenceField, DictField of
ReferenceField, simple ReferenceField.

Supports the import of JSON files containing ids, and json files not containing ids (manual schema definition)

Probably does not support ReferenceFields in EmbeddedDocuments, nor EmbeddedDocumentListFields or equivalent.

Created 3 new unit tests to check the behaviour

### CODE REVIEW

File: mongoengine_goodjson/utils.py [hunk 1]
```diff
@@ -30,7 +30,11 @@ def wrapper(*args, **kw):
 @singledispatch
 def normalize_reference(ref_id, fld):
     """Normalize Reference."""
-    return ref_id and fld.to_python(ref_id) or None
+    try:
+        fld_to_python = fld.to_python(ref_id)
+    except AttributeError:
+        fld_to_python = fld
```
[the remark this answers is not in the document]
Comment (EloiZalczer) [reply mongoengine_goodjson/utils.py]: That works, however I think it hurts the readability. If you think this is better, please go ahead and implement it.

File: tests/integration/fixtures/reference.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ def generate_test_data(cls, additional_suffix=""):
         )
 
 
-class ExtraReference(Dictable, db.Document):
+class ExtraReference(Dictable, gj.Document):
```
[the remark this answers is not in the document]
Comment (EloiZalczer) [reply tests/integration/fixtures/reference.py]: I have found this problem when writing the test_deserialize_without_id test. Basically, the error occurs when the document is loaded from a JSON without ids and saved (meaning that ids are created on the fly). If ExtraReference is not an instance of gj.Document, all ExtraReference objects are not saved to the database.

I am not sure what causes this, I had understood it a few weeks ago when I worked on it but after the holidays I don't remember. There is probably some behavior that is overridden by goodjson and causes this problem. Maybe with a fresh eye and some investigation, you will find the issue! If you change it back to db.Document, you will see the test fail.

File: mongoengine_goodjson/document.py [hunk 3]
```diff
@@ -284,19 +287,99 @@ def from_json(cls, json_str, created=False, *args, **kwargs):
...
+        # manually
+        from_son_result._changed_fields = list(cls._fields.keys())
+
+        if atLeastOneReference:
+            # If the document contains at least one reference, override the
+            # save() method to save the referenced documents at the same time as
+            # the master document. Otherwise, the referenced documents would
+            # not be saved and the document would not be valid anymore after a
+            # save and load from the database.
+            def save(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (EloiZalczer) [reply mongoengine_goodjson/document.py]: That could work, but I don't really see an advantage compared to the current implementation? If anything, I think it makes it less readable. If you think this is better, please go ahead and implement it :)

File: mongoengine_goodjson/document.py [hunk 4]
```diff
@@ -284,19 +287,99 @@ def from_json(cls, json_str, created=False, *args, **kwargs):
...
+                for fldname, fld in cls._fields.items():
+                    if isinstance(fld, db.ReferenceField) or \
+                            isinstance(fld, FollowReferenceField):
```
[the remark this answers is not in the document]
Comment (EloiZalczer) [reply mongoengine_goodjson/document.py]: Done

File: mongoengine_goodjson/document.py [hunk 5]
```diff
@@ -284,19 +287,99 @@ def from_json(cls, json_str, created=False, *args, **kwargs):
...
+                        getattr(self, fldname).save(**kwargs)
+                    elif isinstance(fld, db.DictField) and \
+                        (isinstance(fld.field, db.ReferenceField) or
+                            isinstance(fld.field, FollowReferenceField)):
```
[the remark this answers is not in the document]
Comment (EloiZalczer) [reply mongoengine_goodjson/document.py]: Done

File: mongoengine_goodjson/document.py [hunk 6]
```diff
@@ -284,19 +287,99 @@ def from_json(cls, json_str, created=False, *args, **kwargs):
...
+                        field = getattr(self, fldname)
+                        for key, value in field.items():
+                            field[key].save(**kwargs)
+                    elif isinstance(fld, db.ListField) and \
+                        (isinstance(fld.field, db.ReferenceField) or
+                            isinstance(fld.field, FollowReferenceField)):
```
[the remark this answers is not in the document]
Comment (EloiZalczer) [reply mongoengine_goodjson/document.py]: Done

File: mongoengine_goodjson/document.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hiroaki-yamamoto) [reply mongoengine_goodjson/document.py]: Readability is the current problem to this project as you may know, and ok. I accept it.

File: mongoengine_goodjson/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hiroaki-yamamoto) [reply mongoengine_goodjson/utils.py]: Well, using try-except increases code metrics (Check [here](https://radon.readthedocs.io/en/latest/intro.html#cyclomatic-complexity)) (iow, complexity of the code). I think the code get complicated when the metrics gets increased and this code seems to be redundant a little.

File: tests/integration/fixtures/reference.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hiroaki-yamamoto) [reply tests/integration/fixtures/reference.py]: Ok, I will check how it "fail" after merge. :+1:

File: mongoengine_goodjson/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EloiZalczer) [reply mongoengine_goodjson/utils.py]: Alright, done.

**Comment by hiroaki-yamamoto** [comment]
Could you fix the Lint?
> mongoengine_goodjson/document.py:308:80: E501 line too long (131 > 79 characters)
> mongoengine_goodjson/document.py:324:80: E501 line too long (122 > 79 characters)
> mongoengine_goodjson/document.py:343:80: E501 line too long (118 > 79 characters)
> mongoengine_goodjson/document.py:347:80: E501 line too long (80 > 79 characters)
> mongoengine_goodjson/document.py:358:80: E501 line too long (80 > 79 characters)
> mongoengine_goodjson/document.py:364:80: E501 line too long (82 > 79 characters)
> mongoengine_goodjson/document.py:367:80: E501 line too long (103 > 79 characters)
> tests/integration/document/test_follow_reference.py:63:80: E501 line too long (89 > 79 characters)

**Comment by EloiZalczer** [comment]
Lint is corrected, as well as a compatibility problem with python 2.7 (I don't know if you still want to maintain compatibility?)

However CircleCI still doesn't pass because of this error:
Coverage.py warning: Couldn't read data from '/root/ci/.coverage.3': UnicodeDecodeError: 'utf-8' codec can't decode byte 0x96 in position 106: invalid start byte
Coverage.py warning: Couldn't read data from '/root/ci/.coverage.2': UnicodeDecodeError: 'utf-8' codec can't decode byte 0x96 in position 106: invalid start byte
No usable data files

I think this happens in files that are generated in CircleCI and I am not familiar with it so I'll let you look into it.

**Pull request merged**